lean/main/dsls.lean

```diff
@@ -0,0 +1,514 @@
+
+We shall fix this omission, and finally parse the addition node. 
+We go directly to parsing with `elab`, since the `macro_rules`
+approach does not have anything more interesting to say.
+-/
+
+def elab_aexp_plus (s: Syntax): TermElabM Expr := 
+match s with 
+| `(imp_aexp| $x:imp_aexp + $y:imp_aexp) => do 
+     -- recrsively expand xExpr, yExpr via `Term.elabTerm`
+     let xExpr <- Term.elabTerm (<- `([imp_aexp'| $x])) (expectedType? := none)
+     let yExpr <- Term.elabTerm (<- `([imp_aexp'| $y])) (expectedType? := none)
+     mkAppM ``AExp.APlus #[xExpr, yExpr]
+| _ => do 
+   dbg_trace "elab_aexp_plus failed"
+   throwUnsupportedSyntax
+/-
+Note that we make use of a couple of features here:
+1. `TermElabM` allows us to use macros, so we build the
+    piece of syntax `[imp_aexp'| $x]`. We can see that `TermElabM`
+    in this way already subsumes `macroM`.
+2. To recursively expand `imp_aexp`, we use `Term.elabTerm`.  
+3. See that there is an option, `expectedType? : Option Expr` which allows
+   us to declare the type we expect the elaborated term to be. We ignore
+   this in the tutorial, but make a point to note it, since this shows that
+   `TermElabM` really is capable of reasoning about _types_, and not just
+   `Syntax`.
+-/
+
+elab "[imp_aexp'|" s:imp_aexp "]" : term => elab_aexp_plus s
+def eg_aexp_plus_elab: AExp := [imp_aexp'| foo + bar]
+#print eg_aexp_plus_elab
+-- AExp.APlus (AExp.AVar "foo") (AExp.AVar "bar")
+
+
+/-
+#### Parsing BExp
+
+We repeat the same process, this time for `BExp`.
+This time, we show a different method to writing the elaboration
+function `elab_bexp: Syntax → TermElabM Expr`, where we write
+the function for all `BExp`s at once. This allows us to
+write it as a regular lean function, and use regular recursion
+to elaborate our `BExp`.
+-/
+
+
+declare_syntax_cat imp_bexp
+syntax ident : imp_bexp
+syntax imp_aexp "<" imp_aexp : imp_bexp
+syntax imp_bexp "&&" imp_bexp : imp_bexp
+
+/-
+We first create a helper to function
+to convert Booleans into `Expr`s.
+-/
+def mkBoolLit: Bool -> Expr
+| true => mkConst ``Bool.true
+| false => mkConst ``Bool.false
+
+
+/-
+We then write the elaborator for `BExp` as `elab_bexp`. We first
+handle `true`, `false`, and identifiers in the natural way:
+-/
+
+partial def elab_bexp (s: Syntax): TermElabM Expr := 
+match s with
+| `(imp_bexp| $n:ident) =>
+     let str := n.getId.toString
+     match str with 
+     | "true" => mkAppM ``BExp.BBool #[mkBoolLit true]
+     | "false" => mkAppM ``BExp.BBool #[mkBoolLit false]
+     | n => mkAppM ``BExp.BVar #[mkStrLit str]
+
+/-
+To elaborate the less than (`<`) operator on `aexp`s, we wrie a helper called
+`elab_aexp`, that calls `elabTerm` on the term `[imp_aexp'| $s]`. This produces
+an `Expr` node, which we use to build a `BExp.BLess`
+-/
+| `(imp_bexp| $x:imp_aexp < $y:imp_aexp) => do  
+     let elab_aexp (s: Syntax): TermElabM Expr := do
+        Term.elabTerm (<- `([imp_aexp'| $s])) none
```

Comment:
Same issue as before. Can we do it without `Term.elabTerm`?

Reply:
Same answer as before, I don't think it works without `Term.elabTerm`.